src/LanguageServer/Impl/CodeActions/MissingImportCodeActionProvider.cs

```diff
@@ -0,0 +1,731 @@
+            // for example, numpy.chararry is __all__.extended from numpy.core.chararray and etc.
+            // so we could leave only numpy.chararray and remove ones like numpy.core.chararray and etc. but for now,
+            // we show all those but in certain order so that numpy.chararray shows up top
+            // this heuristic still has issue with something like os.path.join since no one import macpath, macpath join shows up high
+            var sourceDeclarationFullNames = importFullNameMap.Where(kv => kv.Value.Symbol != null)
+                                                              .GroupBy(kv => kv.Value.Symbol.Definition, LocationInfo.FullComparer)
+                                                              .Where(FilterSourceDeclarations)
+                                                              .Select(g => g.First().Key);
+
+            foreach (var fullName in OrderImportNames(sourceDeclarationFullNames)) {
+                importFullNameMap.Remove(fullName);
+                yield return fullName;
+            }
+
+            // put modules that are imported next
+            foreach (var fullName in OrderImportNames(importFullNameMap.Where(FilterImportedModules).Select(kv => kv.Key))) {
+                importFullNameMap.Remove(fullName);
+                yield return fullName;
+            }
+
+            // put members that are imported next
+            foreach (var fullName in OrderImportNames(importFullNameMap.Where(FilterImportedMembers).Select(kv => kv.Key))) {
+                importFullNameMap.Remove(fullName);
+                yield return fullName;
+            }
+
+            // put members whose module is imported next
+            foreach (var fullName in OrderImportNames(importFullNameMap.Where(FilterImportedModuleMembers).Select(kv => kv.Key))) {
+                importFullNameMap.Remove(fullName);
+                yield return fullName;
+            }
+
+            // put things left here.
+            foreach (var fullName in OrderImportNames(importFullNameMap.Select(kv => kv.Key))) {
+                yield return fullName;
+            }
+
+            List<string> OrderImportNames(IEnumerable<string> fullNames) {
+                return fullNames.OrderBy(n => n, ImportNameComparer.Instance).ToList();
+            }
+
+            bool FilterSimpleName(KeyValuePair<string, ImportInfo> kv) => kv.Key.IndexOf(".") < 0;
+            bool FilterImportedMembers(KeyValuePair<string, ImportInfo> kv) => !kv.Value.IsModule && kv.Value.MemberImported;
+            bool FilterImportedModuleMembers(KeyValuePair<string, ImportInfo> kv) => !kv.Value.IsModule && kv.Value.ModuleImported;
+            bool FilterImportedModules(KeyValuePair<string, ImportInfo> kv) => kv.Value.IsModule && kv.Value.MemberImported;
+
+            bool FilterSourceDeclarations(IGrouping<LocationInfo, KeyValuePair<string, ImportInfo>> group) {
+                var count = 0;
+                foreach (var entry in group) {
```

Comment:
So what does it do? Looks like it works with first value, if it exists and then hard-return on second? Seems a bit complicated.... Am I missing something

Reply:
there is a comment explaining what it does at the call site (at the top of same method)